### PR TITLE
[Servo] Support single-element joint limit margins vector and fix joint halting logic for multi-DOF Joints

### PR DESCRIPTION
**Pull request by sea-bass**
### Description

This PR allows the `joint_limit_margins` parameter of MoveIt Servo to be a single-element vector (and changes the defaults accordingly, so it works off-the-shelf with every robot).

It also fixes a not-yet-seen issue in which the joint halting logic would have failed if there were multi-DOF Joints. For those who were spammed by this PR, I was making sure this fix was correct, so please give it a look!

Closes https://github.com/moveit/moveit2/issues/2843.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [x] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [x] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/moveit/moveit2/blob/main/MIGRATION.md) notes
- [x] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html)
- [x] Include a screenshot if changing a GUI
- [x] While waiting for someone to review your request, please help review [another open pull request](https://github.com/moveit/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_ros/moveit_servo/src/servo.cpp [hunk 1]
```diff
@@ -364,7 +361,7 @@ KinematicState Servo::haltJoints(const std::vector<int>& joints_to_halt, const K
     // Halt only the joints that are out of bounds
     bounded_state.positions = target_state.positions;
     bounded_state.velocities = target_state.velocities;
-    for (const int idx : joints_to_halt)
+    for (const auto idx : joint_indices_to_halt)
```
Comment (sea-bass) [inline moveit_ros/moveit_servo/src/servo.cpp:364]: Todo: This also needs to handle multi dof properly

**Review by sjahr — COMMENTED** [review]
Nice work! I have some comments but otherwise I think this is good to go soon

File: moveit_ros/moveit_servo/include/moveit_servo/servo.hpp [hunk 2]
```diff
@@ -230,6 +230,9 @@ class Servo
 
   // Map between joint subgroup names and corresponding joint name - move group indices map
   std::unordered_map<std::string, JointNameToMoveGroupIndexMap> joint_name_to_index_maps_;
+
+  // The current joint limit safety margins for each active joint position variable.
+  std::vector<double> joint_limit_margins_;
```
[the remark this answers is not in the document]
Comment (sea-bass) [reply moveit_ros/moveit_servo/include/moveit_servo/servo.hpp:235]: Does this work? I would prefer to do that too, but didn't think we should be messing with the parameters that way.

File: moveit_ros/moveit_servo/src/servo.cpp (hunk 1 above)
Comment (sea-bass) [reply moveit_ros/moveit_servo/src/servo.cpp:364]: I will fix it in this PR, else it's still a bug

File: moveit_ros/moveit_servo/include/moveit_servo/servo.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjahr) [reply moveit_ros/moveit_servo/include/moveit_servo/servo.hpp:235]: Not 100% sure about the generate parameters internals but my understanding is that we wouldn't change the ROS parameters but just the local copy owned by the class. This is not supposed to be a blocking comment so, if you think it's a bad idea, we can go ahead and merge this

File: moveit_ros/moveit_servo/src/servo.cpp [hunk 3]
```diff
@@ -247,30 +247,27 @@ bool Servo::validateParams(const servo::Params& servo_params) const
...
-        "The parameter 'joint_limit_margins' must have the same number of elements as the number of joints in the "
-        "move group.  The size of 'joint_limit_margins' is '"
-            << servo_params.joint_limit_margins.size() << "' but the number of joints of the move group '"
-            << servo_params.move_group_name << "' is '"
-            << robot_state->getJointModelGroup(servo_params.move_group_name)->getActiveVariableCount() << "'"
-            << check_yaml_string);
 
-    params_valid = false;
+  const auto num_dofs = robot_state->getJointModelGroup(servo_params.move_group_name)->getActiveVariableCount();
+  if (servo_params.joint_limit_margins.size() == 1u)
```
[the remark this answers is not in the document]
Comment (sea-bass) [reply moveit_ros/moveit_servo/src/servo.cpp:252]: There's another standing issue to even move this param out the loop, so I think I'll choose to not touch this function right now.

File: moveit_ros/moveit_servo/include/moveit_servo/servo.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (sea-bass) [reply moveit_ros/moveit_servo/include/moveit_servo/servo.hpp:235]: Yeah, if it did work I'd have to make those params non-const, so I think it's a bit risky. Might leave it.

File: moveit_ros/moveit_servo/src/utils/common.cpp [hunk 4]
```diff
@@ -408,24 +408,49 @@ double jointLimitVelocityScalingFactor(const Eigen::VectorXd& velocities,
...
+std::vector<size_t> jointVariablesToHalt(const Eigen::VectorXd& positions, const Eigen::VectorXd& velocities,
+                                         const moveit::core::JointBoundsVector& joint_bounds,
+                                         const std::vector<double>& margins)
 {
-  std::vector<int> joint_idxs_to_halt;
-  for (size_t i = 0; i < joint_bounds.size(); i++)
+  std::vector<size_t> variable_indices_to_halt;
+
+  // Now get the scaling factor from joint velocity limits.
+  size_t joint_idx = 0;
```
Comment (sea-bass) [inline moveit_ros/moveit_servo/src/utils/common.cpp:418]:
```suggestion
```

**Comment by sea-bass** [comment]
Will fix this later. Tried to make clang tidy happy from my phone and ... yeah

**Review by sjahr — APPROVED** [review]

**Pull request merged**